### PR TITLE
Differentiate underscore alias from named aliases

### PR DESCRIPTION
**Pull request by zombiefungus**
pre for Fixing Issue 2736 
edited to avoid autoclosing the issue

### CODE REVIEW

File: crates/ra_hir_def/src/nameres/raw.rs [hunk 1]
```diff
@@ -145,14 +145,21 @@ impl_arena_id!(Import);
...
     pub(super) is_glob: bool,
     pub(super) is_prelude: bool,
     pub(super) is_extern_crate: bool,
     pub(super) is_macro_use: bool,
     pub(super) visibility: RawVisibility,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ImportAlias {
+    NoAlias,
```
Comment (Veetaha) [inline crates/ra_hir_def/src/nameres/raw.rs]: Could you please put comments near `NoAlias` and `Alias(Name)` like you did with `Unnamed`? I imagine I wouldn't be able to understand what `NoAlias` means without the context of this PR...

Comment (zombiefungus) [reply crates/ra_hir_def/src/nameres/raw.rs]: Sure, makes total sense

Comment (kjeremy) [reply crates/ra_hir_def/src/nameres/raw.rs]: Can you make them doc comments for each variant? Then they will show up in the IDE.

File: crates/ra_hir_def/src/nameres/raw.rs [hunk 2]
```diff
@@ -145,14 +145,21 @@ impl_arena_id!(Import);
...
+    Unnamed, // use Foo as _;
```
Comment (kjeremy) [inline crates/ra_hir_def/src/nameres/raw.rs]: Make this a doc comment.

File: crates/ra_hir_def/src/nameres/raw.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir_def/src/nameres/raw.rs]: Maybe it could be called `OriginalName` or `Default` :thinking:

Comment (matklad) [reply crates/ra_hir_def/src/nameres/raw.rs]: I suggest

```
enum ImportAlias { Name(Name), Underscore }
```

and using `Option` for no-alias case

File: crates/ra_hir_def/src/path.rs [hunk 3]
```diff
@@ -57,7 +57,12 @@ impl ModPath {
     pub(crate) fn expand_use_item(
         item_src: InFile<ast::UseItem>,
         hygiene: &Hygiene,
-        mut cb: impl FnMut(ModPath, &ast::UseTree, /* is_glob */ bool, Option<Name>),
+        mut cb: impl FnMut(
+            ModPath,
+            &ast::UseTree,
+            /* is_glob */ bool,
+            crate::nameres::raw::ImportAlias,
```
Comment (matklad) [inline crates/ra_hir_def/src/path.rs]: We should move `enum ImportAlias` to this module, and `use` it from `raw`.

**Comment by matklad** [comment]
:+1: 

I think it makes sense to merge this (with fixes applied) even before incorporating into the name resolution, as lowering part is pretty self-contained.

**Comment by kjeremy** [comment]
I'd like to see some tests too :)

**Comment by matklad** [comment]
I am ok merging lowering part without tests, as we don't have tests that
target lowering specifically

**Pull request merged**